### PR TITLE
#32 - Refactor GithubWebhookService

### LINKED ISSUE #32: Refactor GithubWebhookService
Pass only needed parameters to functions instead of collection

### PR DESCRIPTION
**Pull request by orlowski11**
This should close #32

Refactored GithubWebhookService and changed overall structure of webhook handling

### CODE REVIEW

File: app/Services/GithubWebhookService.php [hunk 1]
```diff
@@ -4,44 +4,30 @@
...
 
-    protected function getActionType(Collection $data): string
-    {
-        return $data["action"];
+    public function removeMember(
+        int $organizationId,
+        int $memberId,
+    ): void {
+        $user = User::query()->where("github_id", $memberId)->first();
+        $organization = Organization::query()->where("github_id", $organizationId)->first();
```
Comment (kamilpiech97) [inline app/Services/GithubWebhookService.php]: What about if user does not exists? or oragnization?

File: database/factories/OrganizationFactory.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+use Illuminate\Database\Eloquent\Factories\Factory;
+use Illuminate\Support\Str;
+
+class OrganizationFactory extends Factory
+{
+    public function definition(): array
+    {
+        return [
+            "name" => fake()->name(),
+            "github_id" => rand(),
```
Comment (kamilpiech97) [inline database/factories/OrganizationFactory.php]: Lets use faker method

File: database/factories/UserFactory.php [hunk 3]
```diff
@@ -4,30 +4,22 @@
...
 class UserFactory extends Factory
 {
     public function definition(): array
     {
         return [
             "name" => fake()->name(),
             "email" => fake()->unique()->safeEmail(),
-            "email_verified_at" => now(),
-            "password" => "$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi",
+            "github_id" => rand(),
```
Comment (kamilpiech97) [inline database/factories/UserFactory.php]: Same here

File: tests/Feature/GithubWebhookControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            ],
+        ]);
+
+        $webhookService->shouldReceive("createOrganization")
+            ->once()
+            ->with("test", 123, "http://example.com/avatar.png");
+
+        $controller($request);
+
+        $this->assertTrue(true);
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: Is it needed?

File: tests/Feature/GithubWebhookControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    "avatar_url" => "http://example.com/avatar.png",
+                ],
+            ],
+        ]);
+
+        $webhookService->shouldNotReceive("createOrganization");
+
+        $controller($request);
+
+        $this->assertTrue(true);
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: Is it needed?

File: tests/Feature/GithubWebhookControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            ],
+        ]);
+
+        $webhookService->shouldReceive("removeMember")
+            ->once()
+            ->with(123, 456);
+
+        $controller($request);
+
+        $this->assertTrue(true);
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: Same here?

File: tests/Feature/GithubWebhookControllerTest.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    }
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: Lets create test which send specific request on our API endpoint with payload and check assigning user to the specific organization.

Comment (kamilpiech97) [reply tests/Feature/GithubWebhookControllerTest.php]: Same with other tests.

File: tests/Unit/RemoveMemberTest.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+class RemoveMemberTest extends TestCase
+{
+    use RefreshDatabase;
+
+    public function testRemoveMember(): void
+    {
+        $user = User::factory()->create(["github_id" => 123]);
+        $organization = Organization::factory()->create(["github_id" => 456]);
+
```
Comment (kamilpiech97) [inline tests/Unit/RemoveMemberTest.php:31]:
```suggestion
```
For me its first step of test - preparing data.

File: .env.ci [hunk 9]
```diff
@@ -2,11 +2,18 @@ APP_NAME=gha-analyzer
 APP_ENV=testing
 APP_KEY=base64:WVG2VcHt+a2uzCw7n1tVUBPyw2kof644d6EoYA/UUAg=
 APP_DEBUG=false
-APP_URL=http://gha-analyzer.blumilk.localhost
+APP_URL=http://127.0.0.1
```
Comment (kamilpiech97) [inline .env.ci:5]: hm? is it ok?

File: .env.ci [hunk 10]
```diff
@@ -2,11 +2,18 @@ APP_NAME=gha-analyzer
...
 
 LOG_CHANNEL=stack
 LOG_LEVEL=debug
 
+DB_CONNECTION=pgsql
+DB_HOST=127.0.0.1
```
Comment (kamilpiech97) [inline .env.ci:11]: not docker compoer service name?

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: .env.ci (hunk 10 above)
Comment (krzysztofrewak) [reply .env.ci:11]: Not in CI.

Comment (kamilpiech97) [reply .env.ci:11]: sure my fault -  i did not see filename :facepalm:

File: .env.ci (hunk 9 above)
Comment (kamilpiech97) [reply .env.ci:5]: I did not see filename 🤦, its ok

**Review by EwelinaSkrzypacz — COMMENTED** [review]
Checks failed :(

File: tests/Unit/RemoveMemberTest.php (hunk 8 above)
Comment (orlowski11) [reply tests/Unit/RemoveMemberTest.php:31]: moved it to setup

File: tests/Feature/GithubWebhookControllerTest.php (hunk 7 above)
Comment (orlowski11) [reply tests/Feature/GithubWebhookControllerTest.php]: created requests tests

File: tests/Feature/GithubWebhookControllerTest.php (hunk 4 above)
Comment (orlowski11) [reply tests/Feature/GithubWebhookControllerTest.php]: removed it

File: app/Services/GithubWebhookService.php (hunk 1 above)
Comment (orlowski11) [reply app/Services/GithubWebhookService.php]: added an exception

File: database/factories/OrganizationFactory.php (hunk 2 above)
Comment (orlowski11) [reply database/factories/OrganizationFactory.php]: changed to faker methods

File: database/factories/UserFactory.php (hunk 3 above)
Comment (orlowski11) [reply database/factories/UserFactory.php]: changed to faker methods

File: .env.ci [hunk 11]
```diff
@@ -2,15 +2,24 @@ APP_NAME=gha-analyzer
...
+
 BROADCAST_DRIVER=log
 CACHE_DRIVER=array
 FILESYSTEM_DISK=local
 QUEUE_CONNECTION=sync
 SESSION_DRIVER=array
 SESSION_LIFETIME=120
 MAIL_MAILER=array
+
+GITHUB_WEBHOOK_SECRET=123
```
Comment (EwelinaSkrzypacz) [inline .env.ci]:
```suggestion
GITHUB_WEBHOOK_SECRET=123

```

File: app/Services/GithubWebhookService.php [hunk 12]
```diff
@@ -4,44 +4,39 @@
...
+        int $memberId,
+    ): void {
+        $user = User::query()->where("github_id", $memberId)->first();
 
-                break;
-            case "member_removed":
-                $organization = OrganizationDTO::createFromArray($data["organization"]);
-                $member = MemberDTO::createFromArray($data["membership"]["user"]);
+        if (!$user) {
+            throw new \InvalidArgumentException("User with GitHub ID {$memberId} not found.");
```
Comment (kamilpiech97) [inline app/Services/GithubWebhookService.php]: Is it good exception? Maybe create new one?

File: app/Services/GithubWebhookService.php [hunk 13]
```diff
@@ -4,44 +4,39 @@
...
 
-                $this->organizationService->removeMember(
-                    $member,
-                    $organization,
-                );
+        $organization = Organization::query()->where("github_id", $organizationId)->first();
 
-                break;
+        if (!$organization) {
+            throw new \InvalidArgumentException("Organization with GitHub ID {$organizationId} not found.");
```
Comment (kamilpiech97) [inline app/Services/GithubWebhookService.php]: same here

File: tests/Feature/GithubWebhookControllerTest.php [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+                ],
+            ],
+        ]);
+
+        $this->webhookService->shouldReceive("removeMember")
+            ->once()
+            ->with(123, 456);
+
+        $this->webhookController->__invoke($request);
+    }
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: I am thinking about these tests. I do not think so there are needed if we have the same tests below but I do not know.

Comment (kamilpiech97) [reply tests/Feature/GithubWebhookControllerTest.php]: If u want test things like this:
```
$this->webhookService->shouldReceive("createOrganization")
            ->once()
            ->with("test", 123, "http://example.com/avatar.png");
```
I think, u can do that in tests below.

File: tests/Feature/GithubWebhookControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+
+        $this->assertDatabaseMissing("user_organization", [
+            "organization_id" => $this->organization->id,
+            "user_id" => $this->user->id,
+        ]);
+    }
+
+    public function testRemoveMemberRequestWithWrongUser(): void
+    {
+        $payload = [
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php:149]: U can test here which exception where threw.

Comment (kamilpiech97) [reply tests/Feature/GithubWebhookControllerTest.php:149]: https://backendtea.com/post/phpunit-exception-test/

File: tests/Unit/RemoveMemberTest.php (hunk 8 above)
Comment (kamilpiech97) [reply tests/Unit/RemoveMemberTest.php:31]: if u have one test, u can use `setUp` method but it's not necessary.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: tests/Feature/GithubWebhookControllerTest.php (hunk 15 above)
Comment (orlowski11) [reply tests/Feature/GithubWebhookControllerTest.php:149]: added exception tests

File: tests/Feature/GithubWebhookControllerTest.php (hunk 14 above)
Comment (orlowski11) [reply tests/Feature/GithubWebhookControllerTest.php]: deleted these tests

File: app/Services/GithubWebhookService.php (hunk 12 above)
Comment (orlowski11) [reply app/Services/GithubWebhookService.php]: changed it, now catching and throwing ModelNotFoundException

File: app/Services/GithubWebhookService.php [hunk 16]
```diff
@@ -4,44 +4,35 @@
...
+    public function removeMember(
+        int $organizationId,
+        int $memberId,
+    ): void {
+        try {
+            $user = User::query()->where("github_id", $memberId)->firstOrFail();
+            $organization = Organization::query()->where("github_id", $organizationId)->firstOrFail();
+        } catch (ModelNotFoundException $e) {
+            throw $e;
         }
```
Comment (kamilpiech97) [inline app/Services/GithubWebhookService.php]:
```suggestion
        } catch (ModelNotFoundException $exception) {
            throw $exception;
        }
```

File: app/DTO/MemberDTO.php [hunk 17]
```diff
@@ -1,21 +0,0 @@
...
-
-    public static function createFromArray(array $data): self
-    {
-        $member = new self(
-            $data["id"],
-        );
-
-        return $member;
-    }
-}
```
Comment (kamilpiech97) [inline app/DTO/MemberDTO.php]: Why did you remove DTO?

File: app/DTO/OrganizationDTO.php [hunk 18]
```diff
@@ -1,25 +0,0 @@
...
-{
-    public function __construct(
-        public string $name,
-        public int $githubId,
-        public string $avatarUrl,
-    ) {}
-
-    public static function createFromArray(array $data): self
-    {
-        $organization = new self(
```
Comment (kamilpiech97) [inline app/DTO/OrganizationDTO.php]: Here too?

File: tests/Feature/GithubWebhookControllerTest.php [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+
+class GithubWebhookControllerTest extends TestCase
+{
+    use RefreshDatabase;
+
+    protected $organization;
+    protected $user;
+    protected $organizationName;
+    protected $organizationId;
+    protected $organizationAvatarUrl;
```
Comment (kamilpiech97) [inline tests/Feature/GithubWebhookControllerTest.php]: Lets type that variables

File: app/Services/GithubWebhookService.php [hunk 20]
```diff
@@ -25,16 +26,11 @@ public function removeMember(
...
-
-        $organization = Organization::query()->where("github_id", $organizationId)->first();
-
-        if (!$organization) {
-            throw new \InvalidArgumentException("Organization with GitHub ID {$organizationId} not found.");
+        try {
+            $user = User::query()->where("github_id", $memberId)->firstOrFail();
+            $organization = Organization::query()->where("github_id", $organizationId)->firstOrFail();
+        } catch (ModelNotFoundException $e) {
+            throw $e;
```
Comment (kamilpiech97) [inline app/Services/GithubWebhookService.php]: If u are using `firstOrFail`, u can skip `try catch`

File: app/Http/Controllers/Api/GithubWebhookController.php [hunk 21]
```diff
@@ -16,7 +16,30 @@ public function __construct(
...
+                        $account["avatar_url"],
+                    );
+                }
+
+                break;
+            case "member_removed":
+                $this->webhookService->removeMember
+                (
+                    $request->get("organization")["id"],
+                    $request->get("membership")["user"]["id"],
```
Comment (kamilpiech97) [inline app/Http/Controllers/Api/GithubWebhookController.php]: lets use DTO as previous

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: app/Services/GithubWebhookService.php (hunk 20 above)
Comment (orlowski11) [reply app/Services/GithubWebhookService.php]: removed try catch

File: app/Services/GithubWebhookService.php (hunk 16 above)
Comment (orlowski11) [reply app/Services/GithubWebhookService.php]: removed try catch

File: app/DTO/MemberDTO.php (hunk 17 above)
Comment (orlowski11) [reply app/DTO/MemberDTO.php]: added DTOs

File: app/DTO/OrganizationDTO.php (hunk 18 above)
Comment (orlowski11) [reply app/DTO/OrganizationDTO.php]: added DTOs

File: app/Http/Controllers/Api/GithubWebhookController.php (hunk 21 above)
Comment (orlowski11) [reply app/Http/Controllers/Api/GithubWebhookController.php]: changed it to use DTOs

File: tests/Feature/GithubWebhookControllerTest.php (hunk 19 above)
Comment (orlowski11) [reply tests/Feature/GithubWebhookControllerTest.php]: declared types

**Review by kamilpiech97 — APPROVED** [review]
